### ISSUE TITLE
brod 2.0-dev can't produce to an under-replicated topic

### ISSUE THREAD

**Issue by gg7** [kafka4beam/brod#24]
Setup:
- brod: 2.0-dev / 697bbe8e599a274e810a183a306b5bd43fe597a0
- Kafka: kafka_2.11-0.9.0.0 (2 brokers)
- Zookeeper: 3.4.6 (3 servers)

Topic: "brod", 6 partitions, replication factor of 2.

If both Kafka brokers are up I can produce messages to the topic. However, if one of them goes down I can't:

```
Erlang/OTP 18 [erts-7.1] [source-2882b0c] [64-bit] [smp:8:8] [async-threads:10] [hipe] [kernel-poll:false]

Eshell V7.1  (abort with ^G)
1> f().
ok
2> Hosts = [{"kafka1.vagrant", 9092},{"kafka2.vagrant", 9092}].
[{"kafka1.vagrant",9092},{"kafka2.vagrant",9092}]
3> {ok, ClientPid} =
3>   brod:start_link_client(_ClientId  = brod_client_1,
3>                          _Endpoints = Hosts,
3>                          _Config    = [{restart_delay_seconds, 10}],
3>                          _Producers = [ { <<"brod">>
3>                                         , [ {topic_restart_delay_seconds, 10}
3>                                           , {partition_restart_delay_seconds, 2}
3>                                           , {required_acks, -1}
3>                                           ]
3>                                         }
3>                                       ]).
{ok,<0.38.0>}
4> 
=ERROR REPORT==== 2-Feb-2016::21:53:08 ===
** Generic server <0.44.0> terminating 
** Last message in was {post_init,self,brod_producers_sup,
                           {brod_producers_sup2,brod_client_1,<<"brod">>,
                               [{partition_restart_delay_seconds,2},
                                {required_acks,-1}]}}
** When Server state == {state,undefined,undefined,[],undefined,undefined,
                               undefined,[],undefined,undefined}
** Reason for termination == 
** {{badmatch,{error,9}},
    [{brod_producers_sup,post_init,1,
                         [{file,"src/brod_producers_sup.erl"},{line,93}]},
     {supervisor3,handle_info,2,[{file,"src/supervisor3.erl"},{line,717}]},
     {gen_server,try_dispatch,4,[{file,"gen_server.erl"},{line,615}]},
     {gen_server,handle_msg,5,[{file,"gen_server.erl"},{line,681}]},
     {proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,240}]}]}
```

Trying the same with the "new-consumer-api" branch (4edd0ec610deedddb7fcd65666b7a73d6185b946) gives me a "ReplicaNotAvailable" error:

```
Erlang/OTP 18 [erts-7.1] [source-2882b0c] [64-bit] [smp:8:8] [async-threads:10] [hipe] [kernel-poll:false]

Eshell V7.1  (abort with ^G)
1> f().
ok
2> Hosts = [{"kafka1.vagrant", 9092},{"kafka2.vagrant", 9092}].
[{"kafka1.vagrant",9092},{"kafka2.vagrant",9092}]
3> {ok, ClientPid} =
3>   brod:start_link_client(_ClientId  = brod_client_1,
3>                          _Endpoints = Hosts,
3>                          _Config    = [{restart_delay_seconds, 10}],
3>                          _Producers = [ { <<"brod">>
3>                                         , [ {topic_restart_delay_seconds, 10}
3>                                           , {partition_restart_delay_seconds, 2}
3>                                           , {required_acks, -1}
3>                                           ]
3>                                         }
3>                                       ]).
{ok,<0.38.0>}
4> 
=ERROR REPORT==== 2-Feb-2016::21:54:50 ===
** Generic server <0.44.0> terminating 
** Last message in was {post_init,self,brod_producers_sup,
                           {brod_producers_sup2,brod_client_1,<<"brod">>,
                               [{partition_restart_delay_seconds,2},
                                {required_acks,-1}]}}
** When Server state == {state,undefined,undefined,[],undefined,undefined,
                               undefined,[],undefined,undefined}
** Reason for termination == 
** {{badmatch,{error,'ReplicaNotAvailable'}},
    [{brod_producers_sup,post_init,1,
                         [{file,"src/brod_producers_sup.erl"},{line,93}]},
     {supervisor3,handle_info,2,[{file,"src/supervisor3.erl"},{line,717}]},
     {gen_server,try_dispatch,4,[{file,"gen_server.erl"},{line,615}]},
     {gen_server,handle_msg,5,[{file,"gen_server.erl"},{line,681}]},
     {proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,240}]}]}
```

brod 1.5.3 (742ef99) works.

Terminal session recording: https://asciinema.org/a/0prwuld7v6lwclpizbnn03y6n

Am I not using brod 2.0-dev correctly or is this a bug?

**Comment by id** [comment]
Thanks for the report George!

2.0-dev is not tested as much as current master, so there might be dragons.
We have a running PR for producer, I'll try to find time this week to get it finalized and merged in 2.0-dev. And I will use your scenario to test it.

**Comment by zmstone** [comment]
sorry for the late response, @gg7 
the partition error code 9 is an indicator of partition being under-replicated, rather than a real 'error'.
it should be ignored during producer init phase.

later when producing to such partition, if it's not enough replica for the produce request to proceed, there should be exception returned in produce response.

the fix will be merged to 2.0-dev as a part of pull request #17

**Comment by zmstone** [comment]
btw, start_link_client API changed, here is what I used for testing

```
Hosts = [{"localhost", 9092},{"localhost", 9093}].
{ok, ClientPid} =
brod:start_link_client(_ClientId  = brod_client_1,
_Endpoints = Hosts,
_Producers = [ { <<"brod">>, 
  [ {topic_restart_delay_seconds, 10},
    {partition_restart_delay_seconds, 2},
    {required_acks, -1}]}],
_Consumers=[],
_Config = [{restart_delay_seconds, 10}]).
```

**Comment by zmstone** [comment]
fix has been merged to 2.0-dev.

**Comment by gg7** [comment]
Thanks!
